### PR TITLE
fix(python): collect all connector discovery pages

### PR DESCRIPTION
**Pull request by merttcetn**
# Pull Request Description

## Language / Project Scope

- [ ] TypeScript
- [x] Python
- [ ] Documentation only
- [ ] CI/CD or tooling

## Changes

Python connectors previously consumed only the first page returned by MCP `tools/list`, `resources/list`, and `prompts/list`. Servers that paginate discovery results therefore exposed an incomplete capability set to `MCPClient` and higher-level consumers.

This PR follows each response's `nextCursor` until pagination is exhausted, both during eager connector initialization and when the public discovery methods are called directly. Cursors remain opaque (including an empty string). Malformed pages or cursor cycles emit a warning where appropriate and return the items collected so far instead of failing discovery or looping indefinitely.

## Review Readiness

- [x] The PR is scoped to one issue or one clearly related change
- [x] The PR description explains the user-visible problem and the fix
- [x] The diff avoids unrelated formatting, generated files, and bulk rewrites
- [x] I verified the affected package/docs path with the commands listed below
- [x] I checked for existing open PRs that already solve the same issue

## Implementation Details

1. Add a shared `BaseConnector._collect_paginated()` helper that invokes an MCP list method with successive cursors and combines its item collection.
2. Use the helper for tool, resource, and prompt discovery in base initialization and Streamable HTTP eager initialization.
3. Use the same pagination behavior in the public `list_tools()`, `list_resources()`, and `list_prompts()` methods.
4. Preserve accumulated results for `None`/missing/non-list page collections, accept empty-string cursors, and terminate non-string or repeated cursors safely with warnings.
5. Add unit coverage for multi-page aggregation and malformed pagination edge cases.

No dependencies were added or modified.

---

### Pre-commit Checklist

- [x] Code formatted with `ruff format`
- [x] Linting passes with `ruff check`
- [x] Added or updated tests if needed
- [x] Updated documentation if needed (not required; public API is unchanged)

## Example Usage (Before)

```python
tools = await connector.list_tools()
# Only the first server page was returned when nextCursor was present.
```

## Example Usage (After)

```python
tools = await connector.list_tools()
# Tools from every server page are returned.
```

## Documentation Updates

No documentation changes are required because the public API and return types are unchanged; this corrects protocol behavior.

## Testing

- `pytest -q tests/unit` on Python 3.12 with all extras — 316 passed
- `pytest -q tests/unit/test_http_connector.py` against the minimum supported `mcp==1.24.0` — 32 passed
- `ruff check .` with CI-pinned Ruff 0.15.22 — passed
- `ruff format --check .` with CI-pinned Ruff 0.15.22 — 224 files already formatted

Coverage includes eager HTTP initialization, base connector initialization, direct discovery methods, empty-string cursors, `None`/missing/non-list collections, non-string cursors, and repeated cursor cycles.

## Backwards Compatibility

This is backwards compatible. Single-page servers behave exactly as before; paginated servers now return the complete discovery result.

## Related Issues

No linked issue; found during a connector correctness audit.

### CODE REVIEW

File: libraries/python/mcp_use/client/connectors/base.py [hunk 1]
```diff
@@ -440,6 +437,29 @@ async def _ensure_connected(self) -> None:
...
+        cursor: str | None = None
+        seen_cursors: set[str] = set()
+
+        while True:
+            page = await fetch_page() if cursor is None else await fetch_page(cursor)
+            items.extend(getattr(page, items_attribute))
+
+            next_cursor = getattr(page, "nextCursor", None)
+            if not isinstance(next_cursor, str) or not next_cursor:
+                return items
```
[the remark this answers is not in the document]
Comment (merttcetn) [reply libraries/python/mcp_use/client/connectors/base.py:476]: Addressed in f148255. The helper now returns the accumulated items when a page is None or its collection is missing or None. Focused tests cover all three later-page cases.

File: libraries/python/mcp_use/client/connectors/base.py [hunk 2]
```diff
@@ -440,6 +437,29 @@ async def _ensure_connected(self) -> None:
...
+            if next_cursor in seen_cursors:
+                raise RuntimeError(f"{method_name} returned repeated pagination cursor {next_cursor!r}")
+
+            seen_cursors.add(next_cursor)
+            cursor = next_cursor
```
[the remark this answers is not in the document]
Comment (merttcetn) [reply libraries/python/mcp_use/client/connectors/base.py]: Addressed in f148255. A repeated or cyclic cursor now logs a warning and returns every item collected so far, including items from the terminating page, instead of raising or triggering transport fallback.

File: libraries/python/tests/unit/test_http_connector.py [hunk 3]
```diff
@@ -405,6 +453,51 @@ async def test_list_resources(self, _):
...
+            ListToolsResult(tools=[second_tool]),
+        ]
+
+        result = await self.connector.list_tools()
+
+        self.assertEqual(result, [first_tool, second_tool])
+        self.assertEqual(self.connector._tools, [first_tool, second_tool])
+        self.connector.client_session.list_tools.assert_has_awaits([call(), call("tools-page-2")])
+
+    async def test_list_resources_collects_all_pages(self, _):
```
[the remark this answers is not in the document]
Comment (merttcetn) [reply libraries/python/tests/unit/test_http_connector.py:586]: Added test_list_tools_stops_on_repeated_cursor_preserving_page_items in f148255. It verifies finite await count, exact cursor calls, accumulated results, and the warning.

File: libraries/python/mcp_use/client/connectors/base.py [hunk 4]
```diff
@@ -440,6 +437,29 @@ async def _ensure_connected(self) -> None:
...
+
+        while True:
+            page = await fetch_page() if cursor is None else await fetch_page(cursor)
+            items.extend(getattr(page, items_attribute))
+
+            next_cursor = getattr(page, "nextCursor", None)
+            if not isinstance(next_cursor, str) or not next_cursor:
+                return items
+            if next_cursor in seen_cursors:
+                raise RuntimeError(f"{method_name} returned repeated pagination cursor {next_cursor!r}")
```
[the remark this answers is not in the document]
Comment (merttcetn) [reply libraries/python/mcp_use/client/connectors/base.py]: Addressed in f148255 with explicit coverage for empty-string, None, missing collection, non-string, and repeated cursor paths. The cycle test uses the backward-compatible behavior requested in the other review thread: warn and return collected items rather than raise RuntimeError.

**Comment by merttcetn** [comment]
@pietrozullo This Python PR is ready for maintainer review.

All Copilot/Cubic findings were addressed in `f148255`: paginated discovery now preserves partial results for malformed pages, returns with a warning on cursor cycles, and treats the empty string as a valid opaque cursor. Tests cover empty, `None`, missing/`None` collection, non-string, and repeated cursor behavior. All five review threads are resolved.

Local verification: 315 unit tests passed; the focused connector suite passed 31/31 against minimum `mcp==1.24.0`; CI-pinned Ruff lint/format passed. Fresh CI has green unit (latest/minimum), lint, integration, transport, primitive, and 30/30 + 20/20 conformance checks. The two red report-only jobs successfully collected passing test XML and then received the known fork-token 403 while attempting to post comments.

Could you review when convenient?

**Comment by merttcetn** [comment]
Follow-up: the two new malformed-page findings from Cubic are addressed in `4efcbec`.

- `None`, missing, `None`-valued, and non-list page collections now warn and return the accumulated discovery items safely.
- All 7 review threads are resolved.
- Rebased onto current `upstream/main` (`52710122`).
- Python 3.12 unit suite: 316 passed.
- Minimum supported `mcp==1.24.0` focused suite: 32 passed.
- CI-pinned Ruff lint/format: passed.
- GPT-5.6 Sol/max final audit: no code-level findings.

@pietrozullo this is ready for another maintainer review once the refreshed CI finishes.

**Comment by khandrew1** [comment]
Thanks for taking the time to contribute. Our focus right now is on the TypeScript SDK, so we're limiting Python PRs to fixes for concrete problems that contributors have run into firsthand. This PR doesn't appear to address that kind of issue, so we're closing it for now. If it was prompted by something you hit in practice, please open an issue with the details and a reproduction, and we'll happily reopen this.
